### PR TITLE
Feature/add visualization obstacle

### PR DESCRIPTION
**Pull request by ahaffemayer**

### CODE REVIEW

**Comment by ahaffemayer** [comment]
every feature was tested. LGTM.

File: config/controller_configs.yaml
Comment (Kotochleb) [inline config/controller_configs.yaml:1]: Ar you sure those parameters should changed in this PR?

File: config/obstacle_params.yaml
Comment (Kotochleb) [inline config/obstacle_params.yaml:1]: Should this YAML contain commented out shapes?

File: config/pose_publisher_params.yaml
Comment (Kotochleb) [inline config/pose_publisher_params.yaml:1]: This file should also be modyfied?

File: src/crocoddyl_reaching.cpp
Comment (Kotochleb) [inline src/crocoddyl_reaching.cpp:1]: Is this change related to the whole point of this PR?

File: include/panda_torque_mpc/common.h
Comment (Kotochleb) [inline include/panda_torque_mpc/common.h:1]: I am pretty sure there were no changes in this file and the only thing is autoformatting. Can you discrad changes in here?

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include <Eigen/Core>
+#include <boost/smart_ptr/shared_ptr.hpp>
+#include <chrono>
+#include <geometry_msgs/Pose.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <ratio>
+#include <ros/node_handle.h>
+#include <ros/package.h>
+#include <std_msgs/Header.h>
+#include <vector>
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h:12]:
```suggestion
#include <boost/smart_ptr/shared_ptr.hpp>
#include <vector>
```
Unused imports

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#include <pinocchio/spatial/explog.hpp>
+#include <pinocchio/spatial/motion.hpp>
+#include <pinocchio/spatial/se3.hpp>
+#include <pinocchio/algorithm/geometry.hpp>
+#include <pinocchio/algorithm/joint-configuration.hpp>
+#include <pinocchio/fwd.hpp>
+#include <pinocchio/multibody/model.hpp>
+
+#include <hpp/fcl/collision_object.h>
+#include <hpp/fcl/shape/geometric_shapes.h>
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h]:
```suggestion
#include <boost/smart_ptr/shared_ptr.hpp>
#include <vector>

#include <pinocchio/multibody/geometry.hpp>
#include <pinocchio/multibody/fcl.hpp>
#include <pinocchio/multibody/fwd.hpp>

#include <hpp/fcl/shape/geometric_shapes.h>
```
Unused or incorrect imports

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace panda_torque_mpc {
+boost::shared_ptr<pinocchio::GeometryModel> reduce_capsules_robot(
+    const boost::shared_ptr<pinocchio::GeometryModel> &collision_model) {
+  auto reduced_collision_model =
+      boost::make_shared<pinocchio::GeometryModel>(*collision_model);
+  std::vector<std::string> list_names_capsules;
+
+  for (pinocchio::GeomIndex geom_id = 0;
+       geom_id < (pinocchio::GeomIndex)collision_model->ngeoms; ++geom_id) {
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h]:
```suggestion
  for (size_t geom_id = 0; geom_id < collision_model->ngeoms; geom_id++) {
```
Most of the code uses `i++` instead of `++i`. Additionally `pinocchio::GeomIndex` is exactly `size_t` meaning no casting is needed

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    pinocchio::GeometryObject &geometry_object =
+        collision_model->geometryObjects[geom_id];
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h]:
```suggestion
   auto &geometry_object = collision_model->geometryObjects[geom_id];
```
Can be a single line this way

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    if (dynamic_cast<hpp::fcl::Cylinder*>(geometry_object.geometry.get()) != nullptr) {
+
+      size_t last_underscore_pos = geometry_object.name.rfind('_');
+      std::string geom_name = geometry_object.name;
+      std::string link_name = geom_name.substr(0, last_underscore_pos);
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h]:
```suggestion
      const size_t last_underscore_pos = geometry_object.name.rfind('_');
      const std::string geom_name = geometry_object.name;
      const std::string link_name = geom_name.substr(0, last_underscore_pos);
```

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        } else {
+          capsule_name = link_name + "_1";
+          sphere1_name = link_name + "_4";
+          sphere2_name = link_name + "_5";
+
+        }
+        list_names_capsules.push_back(capsule_name);
+
+        const auto cylinder_geom = static_cast<hpp::fcl::Cylinder*>(geometry_object.geometry.get());
+        auto geometry = pinocchio::GeometryObject::CollisionGeometryPtr(
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h]:
```suggestion
        const auto geometry = pinocchio::GeometryObject::CollisionGeometryPtr(
```

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+            new hpp::fcl::Capsule(cylinder_geom->radius,
+                                  cylinder_geom->halfLength));
+
+        pinocchio::GeometryObject capsule(
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h]:
```suggestion
        const pinocchio::GeometryObject capsule(
```

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    if (dynamic_cast<hpp::fcl::Cylinder*>(geometry_object.geometry.get()) != nullptr) {
+
+      size_t last_underscore_pos = geometry_object.name.rfind('_');
+      std::string geom_name = geometry_object.name;
+      std::string link_name = geom_name.substr(0, last_underscore_pos);
+
+      if ((collision_model->existGeometryName(link_name + "_1") &&
+           collision_model->existGeometryName(link_name + "_2")) ||
+          (collision_model->existGeometryName(link_name + "_4") &&
+           collision_model->existGeometryName(link_name + "_5"))) {
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h]: Comments on what is happening in here are must have`

File: include/panda_torque_mpc/reduce_collision_model.h [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        
+        std::string capsule_name, sphere1_name, sphere2_name;
```
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h:36]: Is it capsule or cylider?

File: nodes/SDFGenerator.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+    def _generate_random_color(self):
+        """
+        Generates a random color from the available colors in Gazebo materials.
+
+        Returns:
+            str: Random color name.
+        """
+        return random.choice(self._available_colors)
+
+    def generate_box(self, name, size):
```
Comment (Kotochleb) [inline nodes/SDFGenerator.py]: Missing docstring

File: nodes/SDFGenerator.py
Comment (Kotochleb) [inline nodes/SDFGenerator.py:1]: No type hinting

File: nodes/SDFGenerator.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+                            <radius>{radius}</radius>
+                        </sphere>
+                    </geometry>
+                </collision>
+            </link>
+        </model>
+        """
+        return sdf
+
+    def generate_cylinder(self, name, radius, halfLength):
```
Comment (Kotochleb) [inline nodes/SDFGenerator.py]: Missing docstring

File: nodes/SDFGenerator.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+        """
+        return sdf
+
+
+if __name__ == "__main__":
+    
+    u = SDFGenerator()
+    name = "obstacle"
+    radius = 1
+    print(u.generate_box(name, (1,1,1), include_collision=True))
```
Comment (Kotochleb) [inline nodes/SDFGenerator.py]: Unused code

File: nodes/SDFGenerator.py
Comment (Kotochleb) [inline nodes/SDFGenerator.py:1]: Current implementation due to reusing of most of the SDF tags is a bit prone to typos, but can stay if you relly don't want to change it

File: include/panda_torque_mpc/reduce_collision_model.h
Comment (Kotochleb) [inline include/panda_torque_mpc/reduce_collision_model.h:1]: Add comment pointing that this code is hardcoded and can only work with your current URDF of tha panda

File: nodes/obstacle_visualizer.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/python3
+
+from typing import List, Dict
+
+import rospy
+
+from std_msgs.msg import Header
+from geometry_msgs.msg import Pose
+from tf.transformations import quaternion_from_euler
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]:
```suggestion
```
Unused import

File: nodes/obstacle_visualizer.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+from geometry_msgs.msg import Point, Quaternion, Vector3
+from visualization_msgs.msg import Marker, MarkerArray
+from std_msgs.msg import ColorRGBA, Header
+from SDFGenerator import SDFGenerator
+
+from gazebo_msgs.srv import SpawnModel, SpawnModelRequest
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]:
```suggestion
from geometry_msgs.msg import Point, Quaternion, Vector3
from gazebo_msgs.srv import SpawnModel, SpawnModelRequest
from visualization_msgs.msg import Marker, MarkerArray
from std_msgs.msg import ColorRGBA, Header

from SDFGenerator import SDFGenerator
```
This order of includes make more sense

File: nodes/obstacle_visualizer.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    def __init__(self, name: str) -> None:
+        self._node_name = name
+        rospy.init_node(self._node_name, anonymous=False)
+
+        obstacle_idx = 1
+
+        if not rospy.has_param("~obstacle1"):
+            rospy.logerr(
+                f"[{rospy.get_name()}] Param obstacle1 not found. No collision avoidance with an external obstacle computed!"
+            )
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]:
```suggestion
            rospy.logerr(
                f"[{rospy.get_name()}] Param obstacle1 not found. "
                "No collision avoidance with an external obstacle computed!"
            )           
```
Splitting in two lines improves readability

File: nodes/obstacle_visualizer.py [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+
+        obstacle_idx = 1
+
+        if not rospy.has_param("~obstacle1"):
+            rospy.logerr(
+                f"[{rospy.get_name()}] Param obstacle1 not found. No collision avoidance with an external obstacle computed!"
+            )        
+
+        self._obstacles_infos = {}
+        self._obstacles = []
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]: Unused variable

File: nodes/obstacle_visualizer.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+        sdf_generator = SDFGenerator()
+        self._spawn_model_requests = []
+        self._markers = MarkerArray()
+        header = Header(frame_id="world", stamp=rospy.Time.now())
+        for iter, (key, obstacle) in enumerate(self._obstacles_infos.items()):
+            
+            pose=self._parse_poses(obstacle)
+ 
+            sp_req = SpawnModelRequest(
+                model_name = key[1:],
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]:
```suggestion
           obstacle_name = key.lstrip("~")
            sp_req = SpawnModelRequest(
                model_name = obstacle_name,
```
This way the code will be more scallable

File: nodes/obstacle_visualizer.py [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+                ns="",
+                id=iter,
+                action=Marker.MODIFY,
+                pose = pose,
+                color=ColorRGBA(r=0.58, g=0.0, b=0.98, a=0.7),
+                lifetime=rospy.Duration(0.1),
+                frame_locked=False,
+            )
+            if obstacle["type"] == "sphere":
+                m.scale = Vector3(**dict(zip("xyz", [obstacle["radius"] * 2] * 3)))
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py:62]: I know you hate this line as much as I love it so feel free to replace it with
```python
d = obstacle["radius"] * 
m.scale = Vector3(x=d, y=d, z=d)
```
Or something similar

File: nodes/obstacle_visualizer.py [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+            if obstacle["type"] == "box":
+                m.scale = Vector3(x = obstacle["x"], y = obstacle["y"], z = obstacle["z"])
+                m.type = Marker.CUBE
+                sp_req.model_xml = sdf_generator.generate_box(key[1:], [obstacle["x"],obstacle["y"], obstacle["z"] ])
+            if obstacle["type"] == "capsule":
+                m.scale = Vector3(x = obstacle["radius"], y = obstacle["radius"], z = obstacle["halfLength"])
+                m.type = Marker.CYLINDER
+                sp_req.model_xml = sdf_generator.generate_capsule(key[1:], obstacle["radius"],obstacle["halfLength"] )              
+            print(f"key: {key[1:]}")
+            print(f"sp_req.model_xml: {sp_req.model_xml}")
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]: I think those prints are no longer needed. They were useful for debug, but I don't think they are anymore

File: nodes/obstacle_visualizer.py [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+
+        rospy.wait_for_service("/gazebo/spawn_urdf_model")
+        try:
+            model_spawner = rospy.ServiceProxy("/gazebo/spawn_sdf_model", SpawnModel)
+            for sp_req in self._spawn_model_requests:
+                resp = model_spawner(sp_req)
+                if not resp.success:
+                    rospy.logerr("spawning model didn't work.")
+        except rospy.ServiceException as e:
+            print("Service call failed: %s" % e)
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]: Add ROS paramter disabling this part of the code so it can be later used on real robot

File: nodes/obstacle_visualizer.py [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+
+        # -------------------------------
+        #   Timers
+        # -------------------------------
+
+        self._publish_frequency = 10
+        self._control_loop = rospy.Timer(
+            rospy.Duration(1.0 / self._publish_frequency),
+            self._publish_markers_cb,
+        )
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]:
```suggestion
        # Publish markers at 10 Hz
        self._control_loop = rospy.Timer(
            rospy.Duration(0.1, self._publish_markers_cb,
        )
```

File: nodes/obstacle_visualizer.py [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+
+        rospy.loginfo(f"[{rospy.get_name()}] Node started")
+
+    def _parse_poses(self, obstacle: Dict[str, List[float]]) -> Pose:
+        return Pose(
+            # dark magic association of message objects' parameters
+            position=Point(**dict(zip("xyz", obstacle["translation"]))),
+            orientation=Quaternion(**dict(zip("xyzw", obstacle["rotation"]))),
+        )
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py:114]: I think pose parser doesn't have to be a separate function in this case and could be inside of `__init__`. But I leave it for you to judge.

File: nodes/obstacle_visualizer.py [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+
+    def _publish_markers_cb(self, kwargs) -> None:
+        print(f"key: {self._markers}")
```
Comment (Kotochleb) [inline nodes/obstacle_visualizer.py]: The print should be removed as it pollutes the logs

File: src/obstacle_params_parser.cpp [hunk 24]
```diff
@@ -68,7 +69,7 @@ void ObstacleParamsParser::addCollisions() {
                      "config. Try to use "
                      "the ones for the shapes desired, such as 'radius' for "
                      "'sphere','x', 'y', 'z' for 'box' or 'halfLength' and "
-                     "'radius' for 'capsule'."
+                     "'radius' for 'cylinder' and 'capsule'."
```
Comment (Kotochleb) [inline src/obstacle_params_parser.cpp:72]: This log is missing for cylinder and capsule...

**Review by Kotochleb — CHANGES_REQUESTED** [review]

File: config/controller_configs.yaml
Comment (ahaffemayer) [reply config/controller_configs.yaml:1]: yes let's not overcomplicate things as it's only changing a typo

File: config/pose_publisher_params.yaml
Comment (ahaffemayer) [reply config/pose_publisher_params.yaml:1]: yes

File: include/panda_torque_mpc/reduce_collision_model.h (hunk 9 above)
Comment (ahaffemayer) [reply include/panda_torque_mpc/reduce_collision_model.h:36]: capsule

File: nodes/obstacle_visualizer.py (hunk 18 above)
Comment (ahaffemayer) [reply nodes/obstacle_visualizer.py:62]: it's fine let's keep the complicated things here so when ppl see this they'll think we're smart

File: nodes/obstacle_visualizer.py (hunk 22 above)
Comment (ahaffemayer) [reply nodes/obstacle_visualizer.py:114]: it's fine to add a bit of readability

File: nodes/obstacle_visualizer.py (hunk 18 above)
Comment (Kotochleb) [reply nodes/obstacle_visualizer.py:62]: Your're smart*

**Review by Kotochleb — APPROVED** [review]
LGTM
